cmd/metal-api/internal/service/machine-service.go

```diff
@@ -942,54 +967,58 @@ func (r machineResource) allocateMachine(request *restful.Request, response *res
 	if requestPayload.UserData != nil {
 		userdata = *requestPayload.UserData
 	}
+	if requestPayload.Networks != nil && len(requestPayload.Networks) <= 0 {
+		return nil, errors.New("network ids cannot be empty")
+	}
 
-	image, err := r.ds.FindImage(requestPayload.ImageID)
-	if checkError(request, response, utils.CurrentFuncName(), err) {
-		return
+	image, err := ds.FindImage(requestPayload.ImageID)
+	if err != nil {
+		return nil, err
 	}
 
-	if !image.HasFeature(metal.ImageFeatureMachine) {
-		if checkError(request, response, utils.CurrentFuncName(), fmt.Errorf("given image is not usable for a machine, features: %s", image.ImageFeatureString())) {
-			return
-		}
+	imageFeatureType := metal.ImageFeatureMachine
+	if role == metal.RoleFirewall {
+		imageFeatureType = metal.ImageFeatureFirewall
 	}
 
-	user, err := r.userGetter.User(request.Request)
-	if checkError(request, response, utils.CurrentFuncName(), err) {
-		return
+	if !image.HasFeature(imageFeatureType) {
+		return nil, fmt.Errorf("given image is not usable for a %s, features: %s", imageFeatureType, image.ImageFeatureString())
 	}
 
-	spec := machineAllocationSpec{
-		Creator:            user.EMail,
-		UUID:               uuid,
-		Name:               name,
-		Description:        description,
-		Hostname:           hostname,
-		ProjectID:          requestPayload.ProjectID,
-		PartitionID:        requestPayload.PartitionID,
-		SizeID:             requestPayload.SizeID,
-		Image:              image,
-		FilesystemLayoutID: requestPayload.FilesystemLayoutID,
-		SSHPubKeys:         requestPayload.SSHPubKeys,
-		UserData:           userdata,
-		Tags:               requestPayload.Tags,
-		Networks:           requestPayload.Networks,
-		IPs:                requestPayload.IPs,
-		HA:                 false,
-		Role:               metal.RoleMachine,
-	}
-
-	m, err := allocateMachine(utils.Logger(request).Sugar(), r.ds, r.ipamer, &spec, r.mdc, r.actor, r.grpcServer)
-	if checkError(request, response, utils.CurrentFuncName(), err) {
-		utils.Logger(request).Sugar().Errorw("machine allocation went wrong", "error", err)
-		return
+	partitionID := requestPayload.PartitionID
+	sizeID := requestPayload.SizeID
+	// Allocation of a specific machine is requested, therefor size and partition are not given, fetch them
+	if uuid != "" {
+		m, err := ds.FindMachineByID(uuid)
+		if err != nil {
+			return nil, fmt.Errorf("uuid given but no machine found with uuid:%s err:%w", uuid, err)
+		}
```

Comment:
Introduces an additional backend call, which I think gets repeated in `findMachineCandidate`. Maybe it makes sense to store the pointer to this machine entity in the allocation spec and we can get rid off UUID plus the additional backend call.

Reply:
Good idea, done